src/test/java/org/elasticsearch/cluster/ClusterStateDiffPublishingTests.java

```diff
@@ -0,0 +1,461 @@
+        previousClusterState = clusterState;
+        clusterState = ClusterState.builder(clusterState).blocks(ClusterBlocks.EMPTY_CLUSTER_BLOCK).incrementVersion().build();
+        publishStateDiffAndWait(nodeA.action, clusterState, previousClusterState);
+
+        // Adding new node - this node should get full cluster state while nodeB should still be getting diffs
+
+        Node nodeC = createPublishClusterStateAction("nodeC", ImmutableSettings.EMPTY, Version.CURRENT, new PublishClusterStateAction.NewClusterStateListener() {
+            private final AtomicInteger updateCount = new AtomicInteger(0);
+
+            @Override
+            public void onNewClusterState(ClusterState clusterState, NewStateProcessed newStateProcessed) {
+                int update = updateCount.getAndIncrement();
+                logger.debug("Got cluster state update [{}], version [{}], guid [{}], from diff [{}]", update, clusterState.version(), clusterState.uuid(), clusterState.fromDiff());
+                switch (update) {
+                    case 0: // first state
+                        assertFalse(clusterState.fromDiff());
+                        break;
+                    case 1: // update settings
+                        assertTrue(clusterState.fromDiff());
+                        assertThat(clusterState.metaData().transientSettings().get("foo"), equalTo("bar"));
+                        break;
+                    case 2: // skipping test
+                        assertFalse(clusterState.fromDiff());
+                        break;
+                    default:
+                        fail("Unexpected cluster state update " + update);
+
+                }
+                newStateProcessed.onNewClusterStateProcessed();
+            }
+        });
+
+        // cluster state update 3 - register node C
+        previousClusterState = clusterState;
+        discoveryNodes = DiscoveryNodes.builder(discoveryNodes).put(nodeC.discoveryNode).build();
+        clusterState = ClusterState.builder(clusterState).nodes(discoveryNodes).incrementVersion().build();
+        publishStateDiffAndWait(nodeA.action, clusterState, previousClusterState);
+
+        // cluster state update 4 - update settings
+        previousClusterState = clusterState;
+        MetaData metaData = MetaData.builder(clusterState.metaData()).transientSettings(ImmutableSettings.settingsBuilder().put("foo", "bar").build()).build();
+        clusterState = ClusterState.builder(clusterState).metaData(metaData).incrementVersion().build();
+        publishStateDiffAndWait(nodeA.action, clusterState, previousClusterState);
+
+        // cluster state update - skipping one version change - should request full cluster state
+        previousClusterState = ClusterState.builder(clusterState).incrementVersion().build();
+        clusterState = ClusterState.builder(clusterState).incrementVersion().build();
+        publishStateDiffAndWait(nodeA.action, clusterState, previousClusterState);
```

Comment:
and we also add a test where we switch the "master" and make the version go back?